gematria/datasets/exegesis_benchmark_lib.cc

```diff
@@ -261,6 +263,69 @@ Expected<BenchmarkCode> ExegesisBenchmark::parseJSONBlock(
   return BenchCode;
 }
 
+Expected<BenchmarkCode> ExegesisBenchmark::processAnnotatedBlock(
+    std::string_view BlockHex, const BlockAnnotations &Annotations) {
+  std::optional<std::vector<uint8_t>> BytesOr =
+      gematria::ParseHexString(BlockHex);
+
+  if (!BytesOr.has_value())
+    return llvm::make_error<StringError>(
+        errc::invalid_argument,
+        "Malformed basic block: invalid hex value " + BlockHex);
+
+  Expected<std::vector<gematria::DisassembledInstruction>> InstructionsOrErr =
+      gematria::DisassembleAllInstructions(
+          *LLVMMCDisassembler, ExegesisState.getInstrInfo(),
+          ExegesisState.getRegInfo(), ExegesisState.getSubtargetInfo(),
+          *LLVMMCInstPrinter, 0, *BytesOr);
+
+  if (!InstructionsOrErr) return InstructionsOrErr.takeError();
+
+  BenchmarkCode BenchmarkConfiguration;
+
+  std::vector<MCInst> Instructions;
+  Instructions.reserve(InstructionsOrErr->size());
+
+  for (const gematria::DisassembledInstruction &Instruction :
+       *InstructionsOrErr)
+    Instructions.push_back(Instruction.mc_inst);
+
+  BenchmarkConfiguration.Key.Instructions = std::move(Instructions);
+
+  BenchmarkConfiguration.Key.RegisterInitialValues.reserve(
+      Annotations.initial_regs.size());
+
+  for (const RegisterAndValue &RegisterValue : Annotations.initial_regs) {
+    struct RegisterValue ValueToAdd = {
+        .Register = RegisterValue.register_index,
+        .Value = APInt(64, RegisterValue.register_value)};
+
+    BenchmarkConfiguration.Key.RegisterInitialValues.push_back(
+        std::move(ValueToAdd));
+  }
+
+  MemoryValue MemVal = {.Value = APInt(64, Annotations.block_contents),
+                        .SizeBytes = Annotations.block_size,
+                        .Index = 0};
+  BenchmarkConfiguration.Key.MemoryValues["MEM"] = std::move(MemVal);
+
+  for (const uintptr_t AccessedBlock : Annotations.accessed_blocks) {
+    MemoryMapping MemMap = {.Address = AccessedBlock, .MemoryValueName = "MEM"};
+
+    BenchmarkConfiguration.Key.MemoryMappings.push_back(std::move(MemMap));
```

Comment:
If we're reserving space in vectors, may as well do so for `MemoryMappings` too.